### ISSUE TITLE
Feature Request: Exclusive Mapping

### ISSUE THREAD

**Issue by RALifeCoach** [canjs/canjs#2354]
As I understand things, in the current implementation of Map.define, the CanJS system runs through all the properties in an object and attempts to find an entry in the Map define looking to see if there is additional information. This feature would work the other way. It would first look to the Map.defines. Properties found there would be handled (either found in the object or created as default). Properties not found in the Map would be completely ignored by the CanJS system.

**Comment by ccummings** [comment]
This can already by done using [define’s wildcard `*`](https://canjs.com/docs/can.Map.prototype.define.html#section_Defaultbehaviors) to make everything `type: ‘*`’ and then define the Objects and Arrays as Observable by making them `Type: can.Map` or `Type: can.List`

Here's a simple example: http://jsbin.com/zatodu/1/edit?html,js,output

**Comment by RALifeCoach** [comment]
@ccummings Passing in a large object with many unused properties still creates a large performance issue. I still feel this is a valid request and have created a PR to address the issue.

**Comment by rjgotten** [comment]
@ccummings 
I'm going to somewhat echo what @RALifeCoach already mentioned; there's a substantial cost involved just with walking through all the properties of an object literal and publishing them on a `Map`.

This is biting one of our apps **hard** with approx. 6 seconds of processing time spent in up-converting properties to observable and then setting up binding logic on those properties in views, even though those properties never, ever change value.

It would help *tremendously* if there were a way to explicitly skip conversion of properties not listed in define and having them explicitly be non-observable (i.e. make the binder logic aware of their non-observability and give it a fast-path)

**Comment by justinbmeyer** [comment]
can-define is the solution for this. Closing.

**Comment by rjgotten** [comment]
@justinbmeyer 
Just so I have this clear:

`DefineMap` subclasses use the static constructor to wire up the defined properties as JS getter/setter that add observability. And then any instances you create and feed data, i.e. `new DefineMap({ foo: 1, bar : 2 }) work by assigning all the values to the instance, which will only kick in observability if that particular property already exists as a JS getter/setter.

And there's a `seal` property which, when explicitly set to `false`, allows for 'expando' style automatic just-in-time definition of JS getter/setter pairs which makes `DefineMap` work like the old "observe every property" behavior in `Map`.

Is that correct?

**Comment by justinbmeyer** [comment]
@rjgotten you have it!

The only thing I'd add to is:

> And there's a seal property which, when explicitly set to false, allows for 'expando' style maps with automatic just-in-time definition of a JS getter/setter for the incoming assignment, which makes DefineMap work like the old "observe every property" behavior in Map.

This only works if you use `defineMap.set("prop", newVal)` to set an expando property.  If you add the property normally like `defineMap.prop = newVal`, this doesn't setup any sort of observable property. 

So if you wanted to assign properties to a map where some were observable and others not, you could do it like:

```js
var MyMap = DefineMap.extend({seal: false},{
  observableProp1: "string",
  observableProp2: "number",
  observableProp3: "observable"
});

var myMap = new MyMap();
Object.assign(myMap, {
  observableProp1: "hi",
  observableProp2: 5,
  observableProp3: {something: "else"},
  nonObservable1: "nothing",
  nonObservable2: 4,
  nonObservable3: "u"
});
```


Yes, this will only be available in 3.0.  You can find the list of breaking changes here: https://github.com/canjs/canjs/issues/2026#issuecomment-214504889 Though once we release `can-jquery`, I'm hoping that an upgrade is a pretty straightforward process.

If this is important for 2.3, it might be possible to make a "global" script that that could work. It's not priority for me until after 3.0 is out.  But I can help show what's needed.  The biggest change is right here: https://github.com/canjs/can-define/blob/master/can-define.js#L411

That needs to be `can.__observe`.

**Comment by rjgotten** [comment]
@justinbmeyer 
Thanks for the clear explanation.
I'll keep this on my radar.
